Source/SpacePark-API/DataAccess/APICollector.cs

```diff
@@ -0,0 +1,309 @@
+                    if (property.Name == "length")
+                        parsedShip.ShipLength = property.Value.ToString().Replace(",", "").Replace(".", "");
+                return parsedShip;
+            }
+
+            throw new Exception("Parse was empty; Is the URL in correct format?");
+        }
+
+        #endregion
+
+        public static Person ParseUser(string name)
+        {
+            name = System.Web.HttpUtility.UrlDecode(name); //Decode URL if it contains spacebar. This is to ensure it is always correct
+            //-no matter what encoding is used. Most common is percent20-encoding, but could also be plus-encoding(so that spacebars-
+            //will be replaced with + signs)
+            var foundUser = false;
+            var user = new Person();
+            for (var i = 1; i <= 9; i++)
+                if (!foundUser)
+                {
+                    var tempUsers =
+                        ReturnUsersFromList("https://swapi.dev/api/people/?page=" + i);
+                    foreach (var s in tempUsers)
+                    {
+                        if (s.Name.ToLower().Replace(" ", "").Replace("-", "") != name.ToLower().Replace(" ","").Replace("-","")) continue;
+                        foundUser = true;
+                        user = s;
+                        break;
+                    }
+                }
+                else
+                {
+                    break;
+                }
+
+            return foundUser ? user : null;
+        }
+
+        public static Person ParseUserAsync(string name) //This contains a threadstart for the private corresponding method
+        {
+            var user = new Person();
+            var thread = new Thread(() => { user = ParseUser(name); });
+            thread.Start();
+            thread.Join(); //By doing join it will wait for the method to finish
+            return user;
+        }
+
+        public static SpaceShip ParseShip(string model)
+        {
+            model = System.Web.HttpUtility.HtmlDecode(model);
+            
+            var foundShip = false; //A check to verify a ship has been found by name
+            var
+                ship = new SpaceShip(); //Had to initialize because code didn't realize it was bound to the foundShip bool and i got build errors.
+            for (var i = 1; i <= 4; i++)
+            {
+                var tempShips =
+                    ReturnSpaceShipsFromList("https://swapi.dev/api/starships/?page=" + i);
+                foreach (var s in tempShips)
+                {
+                    if (s.Model.ToLower().Replace(" ","").Replace("-", "") != model.ToLower().Replace(" ", "").Replace("-","")) continue;
+                    foundShip = true;
+                    ship = s;
+                }
+            }
+
+            if (!foundShip)
+                throw new Exception("Could not find a ship based on " + model + ". Did you enter the name correctly?");
+            return ship;
```

Comment:
First, I couldn'd add the boolean foundShip, and also the making of the new ship since I had already put in a comment. I found these lines of code hard to read because they are doing alot

let's see,  I could have written it like this
```suggestion
                foreach (var ship in tempShips)
                {
                    if (ship.Model.ToLower().Replace(" ","").Replace("-", "") == model.ToLower().Replace(" ", "").Replace("-",""));
                    return ship;
                }
            }

            if (!foundShip)
                throw new Exception("Could not find a ship based on " + model + ". Did you enter the name correctly?");
```

This way as **soon** as you find the ship it breaks you off the loop and returns you that ship, otherwise you just throw the exception, the way I'm suggesting is less code and easier to understand. 

 You should not continue searching once the ship is found which is what your code is doing with the !=, and the continue. Imagine if you have 1million ships to look for and your code found it on the 4th go in the foreach, it would go on until it looked those 1million times even if it had already found it which would make the application slower, the point is to break out of the loop as soon as it finds it. You guys could use Break.. but that would be more code.. 

Also keep in mind that naming a variable ''s'' is so vague, it makes it harder to read the code

Reply:
Really good insight about the loop! That is very true!